scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordser/ProductLike.scala

```diff
@@ -0,0 +1,165 @@
+      case (existingTreeOpt, (tpe, accessorSymbol, tBuf)) =>
+        existingTreeOpt match {
+          case Some(t) =>
+            val lastCmp = freshT("lastCmp")
+            Some(q"""
+              val $lastCmp = $t
+              if($lastCmp != 0) {
+                $lastCmp
+              } else {
+                ${tBuf.compareBinary(inputStreamA, inputStreamB)}
+              }
+              """)
+          case None =>
+            Some(tBuf.compareBinary(inputStreamA, inputStreamB))
+        }
+    }.getOrElse(q"0")
+  }
+  def put(c: Context)(inputStream: c.TermName, element: c.TermName)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val innerElement = freshT("innerElement")
+
+    elementData.foldLeft(q"") {
+      case (existingTree, (tpe, accessorSymbol, tBuf)) =>
+        q"""
+          $existingTree
+          val $innerElement = $element.$accessorSymbol
+          ${tBuf.put(inputStream, innerElement)}
+          """
+    }
+  }
+
+  def length(c: Context)(element: c.Tree)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): LengthTypes[c.type] = {
+    import c.universe._
+
+    val (constSize, dynamicFunctions, maybeLength, noLength) =
+      elementData.foldLeft((0, List[c.Tree](), List[c.Tree](), 0)) {
+        case ((constantLength, dynamicLength, maybeLength, noLength), (tpe, accessorSymbol, tBuf)) =>
+
+          tBuf.length(q"$element.$accessorSymbol") match {
+            case const: ConstantLengthCalculation[_] => (constantLength + const.asInstanceOf[ConstantLengthCalculation[c.type]].toInt, dynamicLength, maybeLength, noLength)
+            case f: FastLengthCalculation[_] => (constantLength, dynamicLength :+ f.asInstanceOf[FastLengthCalculation[c.type]].t, maybeLength, noLength)
+            case m: MaybeLengthCalculation[_] => (constantLength, dynamicLength, maybeLength :+ m.asInstanceOf[MaybeLengthCalculation[c.type]].t, noLength)
+            case _: NoLengthCalculationAvailable[_] => (constantLength, dynamicLength, maybeLength, noLength + 1)
+          }
+      }
+
+    val combinedDynamic = dynamicFunctions.foldLeft(q"""$constSize""") {
+      case (prev, t) =>
+        q"$prev + $t"
+    }
+
+    if (noLength > 0) {
+      NoLengthCalculationAvailable(c)
+    } else {
+      if (maybeLength.size == 0 && dynamicFunctions.size == 0) {
+        ConstantLengthCalculation(c)(constSize)
+      } else {
+        if (maybeLength.size == 0) {
+          FastLengthCalculation(c)(combinedDynamic)
+        } else {
+
+          // Contains an Option[Either[Int, Int]]
+          val combinedMaybe: Tree = maybeLength.tail.foldLeft(maybeLength.head) {
+            case (hOpt, nxtOpt) =>
+              q"""
+              $hOpt.flatMap { last =>
+                $nxtOpt.map { nxt =>
+                  (last, nxt) match {
```

Comment:
seems like we need a type other than Either here, and use `+` method on that type to implement this logic. I don't think it needs to be coded into the macro.